File: stage3/s3_pipe_01.js
// S3_PIPE_01 — pipe()
function pipe(...fns) {
  return function (value) {
    return fns.reduce((acc, fn) => fn(acc), value);
  };
}

// Tests
const add1 = (x) => x + 1;
const double = (x) => x * 2;
const square = (x) => x * x;

const run = pipe(add1, double, square);
console.log(run(2)); // ((2 + 1) * 2)² = 36

File: stage3/C/S3_PIPE_01.py
# S3_PIPE_01.py

def pipe(*fns):
    def run(x):
        for fn in fns:
            x = fn(x)
        return x
    return run


def main():
    f = lambda x: x + 1
    g = lambda x: x * 2
    h = lambda x: x - 3

    p = pipe(f, g, h)
    for n in [0, 1, 10]:
        print(f"pipe(f,g,h)({n}) -> {p(n)}")


if __name__ == "__main__":
    main()
